### PR TITLE
#1317 Special Case Pattern

### PR DESCRIPTION
**Pull request by ignite1771**
This PR implement the e-commerce system referred in the #1317 as the example for Special Case Pattern. 

To begin with, to focus on the pattern itself, I implement the DB and maintenance lock of the e-commerce system by the singleton instance.

**Use Case**:  When a logged-in user purchases a product, different receipt view model will be showed based on different cases:
- If the transaction is successful, the default receipt will be showed.
- If the transaction is failed, the particular receipt of one of the special cases will be showed, instead of just a null object with no meaning.

To sum up, we encapsulate special cases into subclasses that provide special behaviors of the target returned interface.

### CODE REVIEW

File: special-case/etc/special-case.urm.puml [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+ApplicationServicesImpl -->  "-domain" DomainServicesImpl
+Account --+ Db
+ApplicationServicesImpl ..|> ApplicationServices 
+DomainServicesImpl ..|> DomainServices 
+DownForMaintenance ..|> ReceiptViewModel 
+InsufficientFunds ..|> ReceiptViewModel 
+InvalidUser ..|> ReceiptViewModel 
+OutOfStock ..|> ReceiptViewModel 
+ReceiptDto ..|> ReceiptViewModel 
+@enduml
```
Comment (ohbus) [inline special-case/etc/special-case.urm.puml:119]: Add a new line

File: special-case/pom.xml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    <dependencies>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+
+</project>
```
Comment (ohbus) [inline special-case/pom.xml:22]: Add a new line

File: special-case/src/test/java/com/iluwatar/specialcase/AppTest.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+/**
+ * Application test.
+ */
+public class AppTest {
+
+  @Test
+  void shouldExecuteWithoutException() {
+    assertDoesNotThrow(() -> App.main(new String[]{}));
+  }
+}
```
Comment (ohbus) [inline special-case/src/test/java/com/iluwatar/specialcase/AppTest.java:16]: Add a new line

**Review by ohbus — CHANGES_REQUESTED** [review]

File: special-case/src/test/java/com/iluwatar/specialcase/AppTest.java (hunk 3 above)
Comment (ignite1771) [reply special-case/src/test/java/com/iluwatar/specialcase/AppTest.java]: Thanks. Solve it in 43a9fab.

File: special-case/pom.xml (hunk 2 above)
Comment (ignite1771) [reply special-case/pom.xml]: Thanks. Solve it in 43a9fab.

File: special-case/etc/special-case.urm.puml (hunk 1 above)
Comment (ignite1771) [reply special-case/etc/special-case.urm.puml]: Thanks. Solve it in 43a9fab.

**Review by ohbus — APPROVED** [review]

**Comment by ohbus** [comment]
@ignite1771 Thank you for contributing to this repository.

It was a great pull request overall.

**Comment by ohbus** [comment]
@all-contributors please add @ignite1771 for code
